kong/tools/database_cache.lua

```diff
@@ -120,21 +121,42 @@ function _M.all_apis_by_dict_key()
 end
 
 function _M.get_or_set(key, cb)
+  local lock = resty_lock:new("cache_locks", {
+    exptime = 10,
+    timeout = 10
+  })
+
   local value, err
-  -- Try to get
+
+  -- Try to get the value from the cache
+  value = _M.get(key)
+  if value then return value end
+
+  -- The value is missing, acquire a lock
+  local elapsed, err = lock:lock(key)
+  if not elapsed then
+    ngx.log(ngx.ERR, "failed to acquire cache lock: "..err)
```

Comment:
Prefer using a comma:
```lua
ngx.log(ngx.ERR, "failed to acquire cache lock: ", err)
```
It is less expensive to call `ngx.log` in its variadic form rather than performing the string concatenation on the Lua-land side, which is more expansive.